### PR TITLE
Add upgrade apis

### PR DESCRIPTION
**Pull request by wanghaoran1988**

### CODE REVIEW

File: deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            and upgrade window and freeze window
+          properties:
+            desired:
+              description: Specify the desired release
+              properties:
+                channel:
+                  description: Channel we gonna use for upgrades
+                  type: string
+                force:
+                  default: false
```
Comment (mrbarge) [inline deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:48]: When testing, I couldn't create the CRD due to this:
`The CustomResourceDefinition "upgradeconfigs.upgrade.managed.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[desired].properties[force].de
fault: Forbidden: must not be set`

I tried on a minikube (1.18) and an OSD cluster. After a bit of research I [noticed a comment elsewhere](https://github.com/coreos/prometheus-operator/issues/3122#issuecomment-610419546) noting that defaults are only supported in CRDs using apiextensions v1.

Comment (mrbarge) [inline deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:48]: When testing, I couldn't create the CRD due to this:
`The CustomResourceDefinition "upgradeconfigs.upgrade.managed.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[desired].properties[force].default: Forbidden: must not be set`

I tried on a minikube (1.18) and an OSD cluster. After a bit of research I noticed a comment elsewhere noting that defaults are only supported in CRDs using apiextensions v1.

Comment (wanghaoran1988) [reply deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:48]: Yes, the crds are generated from v1beta1(which is default), I will regenerate using v1

File: deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+              and upgrade window and freeze window
+            properties:
+              desired:
+                description: Specify the desired release
+                properties:
+                  channel:
+                    description: Channel we gonna use for upgrades
+                    type: string
+                  force:
+                    default: false
```
Comment (bmeng) [inline deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:48]: Is the default value redundant if it is required?

File: deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                  maximumCount:
+                    description: Maximum customerFreezeWindow can be created
+                    format: int32
+                    minimum: 0
+                    type: integer
+                  maximumDuration:
+                    description: Maximum duration for a customer freeze window
+                    properties:
+                      unitOfMeasure:
+                        description: 'Valid values are: "Day", "week", "Month"'
```
Comment (bmeng) [inline deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:71]: week should be Week?

File: pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	CompleteTime *metav1.Time `json:"completeTime,omitempty"`
+}
+type UpgradeConditionType string
+
+const (
+	UpgradeValidated UpgradeConditionType = "Validated"
+	UpgradeComplete  UpgradeConditionType = "Completed"
```
Comment (bmeng) [inline pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go:34]: UpgradeCompleted

File: pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+type UpgradePhase string
+
+const (
+	UpgradePhaseNew       UpgradePhase = "New"
+	UpgradePhasePending   UpgradePhase = "Pending"
+	UpgradePhaseUpgrading UpgradePhase = "Upgrading"
+	UpgradePhaseUpgraded  UpgradePhase = "Upgraded"
+	UpgradePhaseFaild     UpgradePhase = "Failed"
```
Comment (bmeng) [inline pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go:64]: UpgradePhaseFailed

File: deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml (hunk 2 above)
Comment (wanghaoran1988) [reply deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:48]: I don't think so

File: deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml (hunk 3 above)
Comment (wanghaoran1988) [reply deploy/crds/upgrade.managed.openshift.io_upgradeconfigs_crd.yaml:71]: Good catch

File: pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go (hunk 4 above)
Comment (wanghaoran1988) [reply pkg/apis/upgrade/v1alpha1/upgradeconfig_types.go:34]: I prefer not have the "Upgrade" prefix, We know it's talking about upgrade even don't have it, right?

**Pull request merged**